### PR TITLE
Move precompile task before migration task

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
### Description

Closes https://github.com/openfoodfoundation/ofn-install/issues/891

This change minimises the amount of time between the database migration and switching the old code for the new code, which should minimise errors during deployment.

### CODE REVIEW

**Comment by Matt-Yorkley** [comment]
Hmmm okay I've done a bit of testing and as expected it works fine with a regular deployment but has problems on the first deployment on a newly installed server, stopping at the `load schema if it isn't loaded` task with this fatal error:

```
rake aborted!
LoadError: cannot load such file -- listen
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/bootsnap-1.16.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:17:in `require'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/bootsnap-1.16.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:17:in `require'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/zeitwerk-2.6.11/lib/zeitwerk/kernel.rb:38:in `require'
/home/openfoodnetwork/apps/openfoodnetwork/releases-old/2023-08-15-124451/config/application.rb:162:in `block (2 levels) in <class:Application>'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/activesupport-7.0.7/lib/active_support/callbacks.rb:445:in `instance_exec'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/activesupport-7.0.7/lib/active_support/callbacks.rb:445:in `block in make_lambda'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/activesupport-7.0.7/lib/active_support/callbacks.rb:199:in `block (2 levels) in halting'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/activesupport-7.0.7/lib/active_support/callbacks.rb:687:in `block (2 levels) in default_terminator'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/activesupport-7.0.7/lib/active_support/callbacks.rb:686:in `catch'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/activesupport-7.0.7/lib/active_support/callbacks.rb:686:in `block in default_terminator'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/activesupport-7.0.7/lib/active_support/callbacks.rb:200:in `block in halting'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/activesupport-7.0.7/lib/active_support/callbacks.rb:595:in `block in invoke_before'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/activesupport-7.0.7/lib/active_support/callbacks.rb:595:in `each'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/activesupport-7.0.7/lib/active_support/callbacks.rb:595:in `invoke_before'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/activesupport-7.0.7/lib/active_support/callbacks.rb:106:in `run_callbacks'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/activesupport-7.0.7/lib/active_support/reloader.rb:88:in `prepare!'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/railties-7.0.7/lib/rails/application/finisher.rb:68:in `block in <module:Finisher>'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/railties-7.0.7/lib/rails/initializable.rb:32:in `instance_exec'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/railties-7.0.7/lib/rails/initializable.rb:32:in `run'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/railties-7.0.7/lib/rails/initializable.rb:61:in `block in run_initializers'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/railties-7.0.7/lib/rails/initializable.rb:60:in `run_initializers'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/railties-7.0.7/lib/rails/application.rb:372:in `initialize!'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/railties-7.0.7/lib/rails/railtie.rb:226:in `public_send'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/railties-7.0.7/lib/rails/railtie.rb:226:in `method_missing'
/home/openfoodnetwork/apps/openfoodnetwork/releases-old/2023-08-15-124451/config/environment.rb:5:in `<main>'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/bootsnap-1.16.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/bootsnap-1.16.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/zeitwerk-2.6.11/lib/zeitwerk/kernel.rb:38:in `require'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/railties-7.0.7/lib/rails/application.rb:348:in `require_environment!'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/railties-7.0.7/lib/rails/application.rb:506:in `block in run_tasks_blocks'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/bugsnag-6.26.0/lib/bugsnag/integrations/rake.rb:20:in `execute'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/rake-13.0.6/exe/rake:27:in `<top (required)>'
/home/openfoodnetwork/.rbenv/versions/3.1.4/bin/bundle:25:in `load'
/home/openfoodnetwork/.rbenv/versions/3.1.4/bin/bundle:25:in `<main>'
```

**Comment by Matt-Yorkley** [comment]
Okay, I've done a bit of testing and as predicted, it works fine on regular deployments to existing servers but there are some issues when deploying to a fresh server for the first time, and this kicks in during the precompile step:

```
INFO -- : Writing /home/openfoodnetwork/apps/openfoodnetwork/releases-old/2023-08-15-130817/public/assets/admin_minimal-40f864c398f7f6917c612aa6dc3810c3fc7c89a41420586ecd4ea1eb9487ef13.js
I, [2023-08-15T13:09:16.845500 #75866]  INFO -- : Writing /home/openfoodnetwork/apps/openfoodnetwork/releases-old/2023-08-15-130817/public/assets/admin_minimal-40f864c398f7f6917c612aa6dc3810c3fc7c89a41420586ecd4ea1eb9487ef13.js.gz
I, [2023-08-15T13:09:18.713173 #75866]  INFO -- : Writing /home/openfoodnetwork/apps/openfoodnetwork/releases-old/2023-08-15-130817/public/assets/web/all-1da1cc6fe6a50c600a4552a2cc1a45297623e3d517b0e95219ed51cd7f798342.js
I, [2023-08-15T13:09:18.713264 #75866]  INFO -- : Writing /home/openfoodnetwork/apps/openfoodnetwork/releases-old/2023-08-15-130817/public/assets/web/all-1da1cc6fe6a50c600a4552a2cc1a45297623e3d517b0e95219ed51cd7f798342.js.gz
W, [2023-08-15T13:09:20.114670 #75866]  WARN -- : Removed sourceMappingURL comment for missing asset 'angular-file-upload.js.map' from /home/openfoodnetwork/.gem/ruby/3.1.0/gems/angularjs-file-upload-rails-2.4.1/app/assets/javascripts/angularjs-file-upload.js
rake aborted!
NoMethodError: undefined method `iso' for nil:NilClass

    country.iso
           ^^^^
/home/openfoodnetwork/apps/openfoodnetwork/releases-old/2023-08-15-130817/app/services/default_country.rb:9:in `code'
/home/openfoodnetwork/apps/openfoodnetwork/releases-old/2023-08-15-130817/app/assets/javascripts/darkswarm/services/gmaps_geo.js.erb.coffee:16:in `__tilt_7160'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/tilt-2.1.0/lib/tilt/template.rb:191:in `bind_call'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/tilt-2.1.0/lib/tilt/template.rb:191:in `evaluate'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/tilt-2.1.0/lib/tilt/template.rb:109:in `render'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/legacy_tilt_processor.rb:25:in `call'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/processor_utils.rb:75:in `call_processor'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/processor_utils.rb:57:in `block in call_processors'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/processor_utils.rb:56:in `reverse_each'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/processor_utils.rb:56:in `call_processors'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/loader.rb:134:in `load_from_unloaded'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/loader.rb:60:in `block in load'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/loader.rb:317:in `fetch_asset_from_dependency_cache'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/loader.rb:44:in `load'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/cached_environment.rb:20:in `block in initialize'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/cached_environment.rb:47:in `load'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/bundle.rb:23:in `block in call'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/utils.rb:200:in `dfs'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/bundle.rb:24:in `call'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/processor_utils.rb:75:in `call_processor'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/processor_utils.rb:57:in `block in call_processors'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/processor_utils.rb:56:in `reverse_each'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/processor_utils.rb:56:in `call_processors'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/loader.rb:134:in `load_from_unloaded'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/loader.rb:60:in `block in load'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/loader.rb:317:in `fetch_asset_from_dependency_cache'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/loader.rb:44:in `load'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/cached_environment.rb:20:in `block in initialize'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/cached_environment.rb:47:in `load'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/base.rb:66:in `find_asset'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/base.rb:73:in `find_all_linked_assets'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/manifest.rb:134:in `block in find'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/manifest.rb:133:in `each'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/manifest.rb:133:in `find'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/sprockets/manifest.rb:186:in `compile'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-rails-3.4.2/lib/sprockets/rails/task.rb:67:in `block (3 levels) in define'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-3.7.2/lib/rake/sprocketstask.rb:147:in `with_logger'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/sprockets-rails-3.4.2/lib/sprockets/rails/task.rb:66:in `block (2 levels) in define'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/bugsnag-6.26.0/lib/bugsnag/integrations/rake.rb:20:in `execute'
/home/openfoodnetwork/.gem/ruby/3.1.0/gems/rake-13.0.6/exe/rake:27:in `<top (required)>'
/home/openfoodnetwork/.rbenv/versions/3.1.4/bin/bundle:25:in `load'
/home/openfoodnetwork/.rbenv/versions/3.1.4/bin/bundle:25:in `<main>'
```

**Comment by Matt-Yorkley** [comment]
Alright, I've added a little fix for first deployments and I think it's working as intended :tada:

File: roles/deploy/tasks/prepare_database.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  notify: restart puma
+
+- name: seed database
+  command: bash -lc "bundle exec rake db:seed RAILS_ENV={{ rails_env }}"
+  args:
+    chdir: "{{ build_path }}"
+  tags:
+    - seed
+    - skip_ansible_lint
+  notify: restart puma
```
Comment (mkllnk) [inline roles/deploy/tasks/prepare_database.yml:28]: I think that we can replace these three commands with one:
```
rails db:setup # Creates all databases, loads all schemas, and initializes with the seed
```

There shouldn't be any pending migrations but if there were then the migration step is run after the asset compilation anyway.

**Review by mkllnk — APPROVED** [review]

File: roles/deploy/tasks/prepare_database.yml (hunk 1 above)
Comment (dacook) [reply roles/deploy/tasks/prepare_database.yml:28]: This sounds like a good idea, and will be faster with less ansible commands and one less rake task.

But more importantly, it means less code, which will remove the need for a new file :D

File: roles/deploy/tasks/deploy.yml [hunk 2]
```diff
@@ -114,14 +114,9 @@
   changed_when: False
   tags: rake
 
-- name: load schema if it isn't loaded
-  command: bash -lc "bundle exec rake db:schema:load RAILS_ENV={{ rails_env }} I_AM_SURE=1"
-  args:
-    chdir: "{{ build_path }}"
+- name: prepare database
```
Comment (dacook) [inline roles/deploy/tasks/deploy.yml:117]: When reading the file it's not immediately clear this task is conditional, so I'd add a clue in the task name (like there used to be). My suggestion:

```suggestion
- name: prepare database for the first time
```

**Review by dacook — APPROVED** [review]
Awesome, this solves the problem and should make a big difference, so I technically approve this PR as it is!

But I also think Maikel's suggestion is too good to pass up..

File: roles/deploy/tasks/prepare_database.yml (hunk 1 above)
Comment (Matt-Yorkley) [reply roles/deploy/tasks/prepare_database.yml:28]: Hmmm that sounds nice and simple, but I'm trying to figure out if we can remove that `I_AM_SURE=1` on line 5, or why it's there in the first place. Ironically I am not sure what it does...

Comment (Matt-Yorkley) [reply roles/deploy/tasks/prepare_database.yml:28]: Ah I see it's related to the `rails_safe_tasks` gem :+1:

Comment (mkllnk) [reply roles/deploy/tasks/prepare_database.yml]: It protects from overriding a production db, right? So, we don't actually want that in our deployment. We only want to do safe things during a normal deploy.

**Review by mkllnk — APPROVED** [review]

File: roles/deploy/tasks/prepare_database.yml (hunk 1 above)
Comment (dacook) [reply roles/deploy/tasks/prepare_database.yml]: > We only want to do safe things during a normal deploy.

I think that's a very good point! We do have a safety check beforehand, but it's checking for a specific table, so not foolproof (eg what if we removed the table one day?).

I agree and think that we should remove ` I_AM_SURE=1` from `deploy.yml`, so I think will open a new PR.

Comment (dacook) [reply roles/deploy/tasks/prepare_database.yml]: Indeed, I think the `db:setup` is non-destructive (it only creates if missing). In that case, the `I_AM_SURE=1` confirmation should have no effect. I don't know how the old `rails_safe_tasks` gem handles this, but in any case this isn't dangerous so I won't bother updating it for now.

**Pull request merged**